### ISSUE TITLE
enb-bem-i18n

### ISSUE THREAD

**Issue by Guria** [bem/bem-forum-content-ru#416]
Наблюдаю активность вокруг `enb-bem-i18n`, однако по описаниям технологий не до конца ясно с чем они работают.
Подскажите какой процесс интернационализации ими поддерживается и реально ли его использовать в своих проектах?

**Comment by apsavin** [comment]
Как я понимаю, работа еще не [закончена](https://github.com/bem/bem-core/issues/983).

**Comment by blond** [comment]
@Guria Есть `BEM.I18N` из `bem-bl`: https://github.com/bem/bem-bl/tree/support/2.x/blocks-common/i-bem/__i18n. Он же сейчас [скопирован](https://github.com/bem/bem-core/tree/v2/common.blocks/i-bem/__i18n) как есть в `bem-core`.

Пакет `enb-bem-i18n` сейчас поддерживает сборку `BEM.I18N` из `bem-bl`. Можно применять и к `bem-core`, но надо понимать, что он прокидывается в глобальный скоуп и всё из этого вытекающее. 

Задача про новый `i18n` — https://github.com/bem/bem-core/issues/576 и про его сборку — https://github.com/bem/bem-core/issues/950. Пакет про сборку нового `i18n` — https://github.com/blond/enb-bem-core-i18n. Задача про портирование этого пакета в `enb-bem-i18n` — https://github.com/enb-bem/enb-bem-i18n/issues/23.

**Comment by Guria** [comment]
@blond 
А можно вкратце описать предполагаемый принцип? Где задаются строки, как подключаются, где интерполируются. Какие форматы описания строк.

**Comment by blond** [comment]
Для каждого языка создаётся отдельный файл `<lang>.js` в директориях с суффиксом `i18n`:

```
block/
    block.i18n/
        ru.js
        en.js
```

`ru.js`

В каждом таком файле хранятся строки (keysets).

```js
module.exports = {
    scope: {
        key: "val"
    }
};
```

После сборки `lang-js` файлов будет возможность вызвать функцию `BEM.I18N`:

```
BEM.I18N('scope', 'key') // val
```

Обычно в качестве скоупа используется имя блока:

`block/block.i18n.ru.js`

```js
module.exports = {
    block: {
        hello: "Привет!"
    }
};
```

`block/block.i18n.en.js`

```js
module.exports = {
    block: {
        hello: "Hello!"
    }
};
```

```js
BEM.lang('ru');
BEM.I18N('block', 'hello'); // Привет!

BEM.lang('en');
BEM.I18N('block', 'hello'); // Hello!
```

Вызов `BEM.I18N` можно применять, например, в шаблонах:

```js
block('block').content()(function() {
    return {
        elem: 'tooltip',
        content: BEM.I18N('block', 'tooltip');
    };
});
```

**Comment by apsavin** [comment]
Разве это не описание технологии из bem-bl?

**Comment by blond** [comment]
@apsavin, он, вопрос был про другое?

**Comment by apsavin** [comment]
Спрашивал не я, возможно, и про это)
Я почему-то подумал, что "предполагаемый" - это про новый i18n, который еще в работе.

**Comment by Guria** [comment]
Вопрос был про новую предполагаемую реализацию.
Про существующую технологию тоже было интересно, спасибо.

**Comment by blond** [comment]
Ключи будут храниться точно так же:

```
block/
  block.i18n/
    ru.js
    en.js
```

`ru.js`

```js
module.exports = {
    scope: {
        key: "val"
    }
};
```

Появится возможность добавлять `common` файлы — `i18n.js`. В таком файле и будет храниться ядро — https://github.com/bem/bem-core/blob/issues/576%40v3/common.blocks/i18n/i18n.i18n.js.

После сборки `lang-js` файлов будет возможность получить `i18n` модуль из YModules:

```js
modules.require('i18n', function (i18n) {
    i18n('scope', 'key'); // val
});
```

В BEMHTML `i18n` будет доступно через `this.i18n`:

```js
block('block').content()(function() {
    return {
        elem: 'tooltip',
        content: this.i18n('block', 'tooltip');
    };
});
```

Появится возможность работы с параметризированными ключами:

```js
{
    'scope-1': {
        key: 'val'
    },
    'scope-2': {
        key: function (params, i18n) {
            return i18n(params.scope, params.key);
        }
    }
}
```

```js
modules.require('i18n', function (i18n) {
    i18n('scope-2', 'key', { scope: 'scope-1', key: 'key' }); // val
});
```

**Comment by pavelpower** [comment]
@blond так эта новая технология уже готова?

**Comment by blond** [comment]
@pavelpower В сыром виде, да:

> Задача про новый i18n — bem/bem-core#576 и про его сборку — bem/bem-core#950. Пакет про сборку нового i18n — https://github.com/blond/enb-bem-core-i18n. Задача про портирование этого пакета в enb-bem-i18n — enb-bem/enb-bem-i18n#23.

**Comment by Sergei-b84** [comment]
@blond
>Пакет enb-bem-i18n сейчас поддерживает сборку BEM.I18N из bem-bl. Можно применять и к bem-core, но надо понимать, что он прокидывается в глобальный скоуп и всё из этого вытекающее.

Расскажите, подробнее о вытекающих. На что повлияет? Советовали бы Вы использовать данный модуль в реальных проектах со средней нагрузкой? На данном форуме используется данная технология или что-то другое?

**Comment by tadatuta** [comment]
@Sergei-b84 Да, здесь и на большинстве сервисов Яндекса используется эта технология. Использовать ее можно смело. Причин сомневаться могу придумать две:
1. Засоряется глобальная область видимости.
2. Мы планируем достаточно скоро выпустить новую версию, использующую модульную систему и новое ядро. После чего текущая реализация будет объявлена устаревшей. Но переход на новую версию должен оказаться достаточно простым, так что сильно переживать в любом случае не стоит.

**Comment by devvk** [comment]
Обновите пожалуйста документацию по i18n с примером рабочего конфигурационного файла. Спасибо

**Comment by tadatuta** [comment]
@devvk как будто бы там актуальная информация.
вот на всякий случай пример внедрения в project-stub: https://github.com/bem/project-stub/commit/344de8a78b4e33d2057eb44526f38d07c637051f
